### PR TITLE
remove index_fill side-effect for scalar tensors

### PR DESCRIPTION
**Pull request by nikitaved**
`index_fill` silently promotes zero dim Tensors to 1-dim Tensors. This PR fixes that.
Was:
```
In [1]: import torch

In [2]: x = torch.tensor(1)

In [3]: idx = torch.tensor(0).long()

In [4]: x.dim()
Out[4]: 0

In [5]: x.index_fill(0, idx, -1).dim()
Out[5]: 1

```
Now:
```
In [6]: x.index_fill(0, idx, -1).dim()
Out[6]: 0

```

### CODE REVIEW

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -737,17 +737,15 @@ Tensor & index_fill_(Tensor & self, int64_t dim, const Tensor & index, Scalar so
   }
 
   // Handle the case when `self` is 0-dim
-  if (0 == self.dim()) {
-    self.unsqueeze_(-1);
-  }
```
Comment (nikitaved) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:742]: The source of the problem.

**Comment by ngimel** [comment]
That's a good fix, can you please add a test for it? Would also be good to add similar test for masked_fill and functions that can exhibit similar behavior.

**Review by ngimel — APPROVED** [review]

**Comment by anjali411** [comment]
@nikitaved could you rebase?
